### PR TITLE
Neg n_workers are now the same as zero

### PR DESCRIPTION
**Pull request by Erotemic**
I foolishly passed the kwarg `num_workers=-1` to a `DataLoader` expecting that it would default to serial processing (I correctly recalled that it should be a non-positive number, but failed to remember that it must be zero to get this behavior). 

Normally this would be fine, but the error message that it spat out when I did this was very confusing:

```
AttributeError: 'DataLoaderIter' object has no attribute 'rcvd_idx'
``` 

What made the matter worse is that I recently put a print statement in that file to test something, so I though I had broken pytorch, but I couldn't figure out where the difference was. After spending too much time searching for cached pyc files, I finally realized my mistake. 

I wanted to submit a patch so it would at least spit out a ValueError. However, I think it is a simpler change to have it accept negative numbers. This change simply makes it so setting num_workers to a non-positive number results in the same behavior as zero.

### CODE REVIEW

**Comment by kmike** [comment]
FWIW, scikit-learn and joblib use a different convention: -1 means "use the same number of workers as CPU cores", -2 means "use num_cores-1", etc. From joblib docs:

> If -1 all CPUs are used. If 1 is given, no parallel computing code is used at all, which is useful for debugging. For n_jobs below -1, (n_cpus + 1 + n_jobs) are used. Thus for n_jobs = -2, all CPUs but one are used.

**Comment by apaszke** [comment]
Taking into account, @kmike's comment I feel like it'd be best to raise `ValueError` for negative values.

**Comment by Erotemic** [comment]
Changed it to a `ValueError`

**Review by apaszke — APPROVED** [review]

File: torch/utils/data/dataloader.py [hunk 1]
```diff
@@ -219,6 +219,9 @@ def __init__(self, loader):
             # prime the prefetch loop
             for _ in range(2 * self.num_workers):
                 self._put_indices()
+        elif self.num_workers < 0:
```
Comment (apaszke) [inline torch/utils/data/dataloader.py]: Actually now that I think about it, this check should probably happen when constructing the data loader and not the iterator. Can you please move it there?

**Review by apaszke — APPROVED** [review]

**Comment by apaszke** [comment]
Thank you!

**Pull request merged**